File: app/src/main/java/co2p/droplet/MainActivity.java
package co2p.droplet;

import android.Manifest;
import android.animation.ValueAnimator;
import android.app.Activity;
import android.app.Dialog;
import android.content.Context;
import android.content.pm.PackageManager;
import android.graphics.Color;
import android.location.Location;
import android.location.LocationListener;
import android.location.LocationManager;
import android.os.Bundle;
import android.support.v4.app.ActivityCompat;
import android.view.View;
import android.widget.ProgressBar;
import android.widget.TextView;

import com.android.volley.Request;
import com.android.volley.RequestQueue;
import com.android.volley.toolbox.StringRequest;
import com.android.volley.toolbox.Volley;

import java.util.Arrays;
import java.util.Calendar;

import static co2p.droplet.Color.backgroundColor;

public class MainActivity extends Activity {

    private Calendar calendar = Calendar.getInstance();
    private long updated = 25;
    private TextView temperature_text;
    private TextView location_text;
    private TextView state_text;
    private ProgressBar spinner;
    private View background;
    private float[] currentBackgroundColor = new float[]{0, 0, 0};
    private boolean animating = false;
    private boolean searching = false;
    private Location location;
    public static final int MY_PERMISSIONS_REQUEST_LOCATION = 57;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        temperature_text = (TextView) findViewById(R.id.temperature);
        location_text = (TextView) findViewById(R.id.location);
        state_text = (TextView) findViewById(R.id.state);
        spinner = (ProgressBar) findViewById(R.id.progress);

    }

    public void setLocation(Location l) {
        location = l;
    }

    public void getLocation() {
        if (!searching) {
            searching = true;
            spinner.setVisibility(View.VISIBLE);
            ActivityCompat.requestPermissions(this,
                    new String[]{Manifest.permission.ACCESS_FINE_LOCATION},
                    MY_PERMISSIONS_REQUEST_LOCATION);
            if (ActivityCompat.checkSelfPermission(this,
                    Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED &&
                    ActivityCompat.checkSelfPermission(this,
                            Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
                searching =false;
                getLocation();
            }


            // Acquire a reference to the system Location Manager
            final LocationManager locationManager =
                    (LocationManager) this.getSystemService(Context.LOCATION_SERVICE);

            location = locationManager.getLastKnownLocation(LocationManager.NETWORK_PROVIDER);

            try {
                requestWeather(location);
            } catch (NullPointerException e) {
                location_text.setText(R.string.locationerror);
                temperature_text.setText("???");
                animateBackground(currentBackgroundColor, backgroundColor(50, getBaseContext()), 2000);
            }

            // Define a listener that responds to location updates
            final LocationListener locationListener = new LocationListener() {
                public void onLocationChanged(Location location) {
                    // Called when a new location is found by the network location provider.
                    //locationManager.removeUpdates(this);
                    searching = false;
                    setLocation(location);

                    requestWeather(location);
                }

                public void onStatusChanged(String provider, int status, Bundle extras) {
                }

                public void onProviderEnabled(String provider) {
                }

                public void onProviderDisabled(String provider) {
                }
            };
            locationManager.requestLocationUpdates(LocationManager.NETWORK_PROVIDER, 0, 0, locationListener);
            //Refresh animation
        }

    }

    /* Request updates at startup */
    @Override
    protected void onResume() {
        super.onResume();
        requestWeather(location);
        getLocation();
    }

    /* Remove the locationlistener updates when Activity is paused */
    @Override
    protected void onPause() {
        super.onPause();
    }

    public void onRefresh(View view) {
        requestWeather(location);
        getLocation();
    }

    public void requestWeather(Location location) {
        //TODO fix update time to 10 minutes
        if (location != null && (updated >= calendar.get(Calendar.HOUR_OF_DAY) + 1 || updated == 25)) {
            spinner.setVisibility(View.VISIBLE);

            //weather.getWeather(location.getLatitude(), location.getLongitude());

            RequestQueue request = Volley.newRequestQueue(this);

            String url = "http://api.yr.no/weatherapi/locationforecast/1.9/?lat=" + location.getLatitude() + ";lon=" + location.getLongitude();

            StringRequest stringRequest = new StringRequest(Request.Method.GET, url, response -> {
                XMLParser parser = new XMLParser();
                Weather weather = parser.parse(response);
                updateTemperature(weather);
            }, error -> location_text.setText(R.string.interneterror));

            request.add(stringRequest);

        } else {
            spinner.setVisibility(View.INVISIBLE);
        }

    }

    public void updateTemperature(Weather object) {
        final float[] from = currentBackgroundColor;
        float temp = Float.parseFloat(object.temperature);
        final float[] to = backgroundColor(temp, getBaseContext());

        String location = "";
        background = findViewById(R.id.background);

        //String simpleWeather[] = new String[1];
        updated = calendar.get(Calendar.HOUR_OF_DAY);


        temperature_text.setText(Math.round(temp) + getString(R.string.celcius));
        location_text.setText(location);
        state_text.setText(object.winddirection + ", " + object.windspeed + "m/s");

        animateBackground(from, to, 2000);

        spinner.setVisibility(View.INVISIBLE);

    }

    private void animateBackground(final float[] from, final float[] to, int time) {
        animating = true;
        ValueAnimator anim = ValueAnimator.ofFloat(0, 1);   // animate from 0 to 1
        anim.setDuration(time);                              // for 2000 ms

        final float[] hsv = new float[3];                  // transition color
        anim.addUpdateListener(animation -> {
            // Transition along each axis of HSV (hue, saturation, value)
            hsv[0] = from[0] + (to[0] - from[0]) * animation.getAnimatedFraction();
            hsv[1] = from[1] + (to[1] - from[1]) * animation.getAnimatedFraction();
            hsv[2] = from[2] + (to[2] - from[2]) * animation.getAnimatedFraction();

            background.setBackgroundColor(Color.HSVToColor(hsv));
            if (Arrays.equals(from, to)) {
                animating = false;
            }
        });

        anim.start();
        currentBackgroundColor = to;
    }

    @Override
    public void onStart() {
        super.onStart();
    }

    @Override
    public void onStop() {
        super.onStop();
    }

}

File: app/src/main/java/co2p/droplet/MainActivity.kt
package co2p.droplet

import android.Manifest
import android.animation.ValueAnimator
import android.app.Activity
import android.content.Context
import android.content.pm.PackageManager
import android.graphics.Color
import android.location.Location
import android.location.LocationListener
import android.location.LocationManager
import android.os.Bundle
import android.support.v4.app.ActivityCompat
import android.view.View
import android.widget.ProgressBar
import android.widget.TextView

import com.android.volley.Request
import com.android.volley.toolbox.StringRequest
import com.android.volley.toolbox.Volley

import java.util.Arrays
import java.util.Calendar

import co2p.droplet.Color.backgroundColor
import java.lang.Float

class MainActivity : Activity() {

    private val calendar = Calendar.getInstance()
    private var updated = 25
    private var temperature_text: TextView? = null
    private var location_text: TextView? = null
    private var state_text: TextView? = null
    private var spinner: ProgressBar? = null
    private var background: View? = null
    private var currentBackgroundColor = floatArrayOf(0f, 0f, 0f)
    private var animating = false
    private var searching = false
    private var location: Location? = null


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        temperature_text = findViewById<TextView>(R.id.temperature)
        location_text = findViewById<TextView>(R.id.location)
        state_text = findViewById<TextView>(R.id.state)
        spinner = findViewById<ProgressBar>(R.id.progress)
    }

    fun setLocation(l: Location) {
        location = l
    }

    fun getLocation() {
        if (!searching) {
            searching = true
            spinner!!.visibility = View.VISIBLE
            ActivityCompat.requestPermissions(this,
                    arrayOf(Manifest.permission.ACCESS_FINE_LOCATION),
                    MY_PERMISSIONS_REQUEST_LOCATION)
            if (ActivityCompat.checkSelfPermission(this,
                            Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED && ActivityCompat.checkSelfPermission(this,
                            Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
                searching = false
                getLocation()
            }


            // Acquire a reference to the system Location Manager
            val locationManager = this.getSystemService(Context.LOCATION_SERVICE) as LocationManager

            location = locationManager.getLastKnownLocation(LocationManager.NETWORK_PROVIDER)

            try {
                requestWeather(location)
            } catch (e: NullPointerException) {
                location_text!!.setText(R.string.locationerror)
                temperature_text!!.text = "???"
                animateBackground(currentBackgroundColor, backgroundColor(50f, baseContext), 2000)
            }

            // Define a listener that responds to location updates
            val locationListener = object : LocationListener {
                override fun onLocationChanged(location: Location) {
                    // Called when a new location is found by the network location provider.
                    //locationManager.removeUpdates(this);
                    searching = false
                    setLocation(location)

                    requestWeather(location)
                }

                override fun onStatusChanged(provider: String, status: Int, extras: Bundle) {}

                override fun onProviderEnabled(provider: String) {}

                override fun onProviderDisabled(provider: String) {}
            }
            locationManager.requestLocationUpdates(LocationManager.NETWORK_PROVIDER, 0, 0f, locationListener)
        }
    }

    /* Request updates at startup */
    override fun onResume() {
        super.onResume()
        requestWeather(location)
        getLocation()
    }

    /* Remove the locationlistener updates when Activity is paused */
    override fun onPause() {
        super.onPause()
    }

    fun onRefresh(view: View) {
        requestWeather(location)
        getLocation()
    }

    fun requestWeather(location: Location?) {
        //TODO fix update time to 10 minutes
        location_text!!.text = getString(R.string.loading_weather)

        if (location != null && (updated >= calendar.get(Calendar.HOUR_OF_DAY) + 1 || updated == 25)) {
            spinner!!.visibility = View.VISIBLE

            //weather.getWeather(location.getLatitude(), location.getLongitude());


            val request = Volley.newRequestQueue(this)

            val url = "https://api.met.no/weatherapi/locationforecast/1.9/?lat=" + location.latitude + ";lon=" + location.longitude

            val stringRequest = StringRequest(Request.Method.GET, url, { response ->
                val parser = XMLParser()
                val weather = parser.parse(response)
                updateTemperature(weather)
            }) { error ->
                var statusCode = 0
                try {
                    statusCode = error.networkResponse.statusCode
                    println("StatusCode: " + statusCode)
                } catch (e: NullPointerException) {
                    location_text!!.setText(R.string.airplanemode)
                }
                when (statusCode) {
                    404 -> location_text!!.setText(R.string.interneterror)
                    429 -> location_text!!.setText("Error 429\n Too Many Requests")
                    else -> error.printStackTrace()
                }
                spinner!!.visibility = View.INVISIBLE
            }

            request.add(stringRequest)
            println(url)
        } else {
            spinner!!.visibility = View.INVISIBLE
        }
    }

    fun updateTemperature(weather: Weather) {
        val from = currentBackgroundColor
        val temp = Float.parseFloat(weather.temperature)
        val to = backgroundColor(temp, baseContext)

        val location = weather.area
        println("Loaction: " + location)
        background = findViewById(R.id.background)

        //String simpleWeather[] = new String[1];
        updated = calendar.get(Calendar.HOUR_OF_DAY)

        temperature_text!!.text = Math.round(temp).toString() + getString(R.string.celcius)
        location_text!!.text = location
        state_text!!.text = weather.winddirection + ", " + weather.windspeed + getString(R.string.speed_meters_per_sec)

        animateBackground(from, to, 2000)

        spinner!!.visibility = View.INVISIBLE

    }

    private fun animateBackground(from: FloatArray, to: FloatArray, time: Int) {
        animating = true
        val anim = ValueAnimator.ofFloat(0F, 1F)   // animate from 0 to 1
        anim.duration = time.toLong()                              // for 2000 ms

        val hsv = FloatArray(3)                  // transition color
        anim.addUpdateListener { animation ->
            // Transition along each axis of HSV (hue, saturation, value)
            hsv[0] = from[0] + (to[0] - from[0]) * animation.animatedFraction
            hsv[1] = from[1] + (to[1] - from[1]) * animation.animatedFraction
            hsv[2] = from[2] + (to[2] - from[2]) * animation.animatedFraction

            background!!.setBackgroundColor(Color.HSVToColor(hsv))
            if (Arrays.equals(from, to)) {
                animating = false
            }
        }

        anim.start()
        currentBackgroundColor = to
    }

    companion object {
        const val MY_PERMISSIONS_REQUEST_LOCATION = 57
    }
}
